cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -101,6 +110,60 @@
   }
 
   /**
+   * Migrate schedules in a given namespace from app metadata store to ProgramScheduleStoreDataset
+   *
+   * @param namespaceId  the namespace with schedules to be migrated
+   * @param appMetaStore app metadata store with schedules to be migrated
+   * @return the lexicographically largest namespace id String with schedule migration completed
+   */
+  public String migrateFromAppMetadataStore(NamespaceId namespaceId, Store appMetaStore) {
+    String completedNamespace = getMigrationCompleteNamespace();
+    if (completedNamespace != null && completedNamespace.compareTo(namespaceId.toString()) > 0) {
+      return completedNamespace;
+    }
+    for (ApplicationSpecification appSpec : appMetaStore.getAllApplications(namespaceId)) {
+      ApplicationId appId = namespaceId.app(appSpec.getName(), appSpec.getAppVersion());
+      for (ScheduleSpecification scheduleSpec : appSpec.getSchedules().values()) {
+        String scheduleKey = rowKeyForSchedule(appId, scheduleSpec.getSchedule().getName());
+        ProgramSchedule schedule = Schedulers.toProgramSchedule(appId, scheduleSpec);
+        persistSchedule(scheduleKey, schedule);
```

Comment:
right, originally thought that the check for schedule existence in `addSchedules` is not necessary but it doesn't matter to check